tsfc/driver.py

```diff
@@ -359,19 +359,20 @@ def fn(point_set, deriv=None):
             '''Wrapper function for converting UFL `expression` into GEM expression.
             '''
             # Is exterior derivative required?
-            from ufl.operators import Dx, Dn, div, grad, nabla_div, nabla_grad, curl, \
-                                      rot, diff
-            deriv_dict = {'Dx': Dx, 'Dn': Dn, 'div': div, 'grad': grad, 'nabla_div': nabla_div, \
-                          'nabla_grad': nabla_grad, 'curl': curl, 'rot': rot, 'diff': diff}
+            from ufl.operators import Dn, div, grad, nabla_div, nabla_grad, curl, rot
+            deriv_oper = {'Dn': Dn, 'div': div, 'grad': grad, 'nabla_div': nabla_div,
+                          'nabla_grad': nabla_grad, 'curl': curl, 'rot': rot}
             config = kernel_cfg.copy()
             config.update(point_set=point_set)
             if deriv is None:
```

Comment:
If  you make the default value of `deriv` 0 instead of `None` then we can get rid of this whole if statement and just keep lines 372 and 373. Also, abbreviations are the enemy, just call it `derivative`.

Reply:
Wouldn't just doing `grad` require more processing on the FInAT end, if say one were to use a finite element with curls or divs? Though it could be more useful if one wants to use another form language?

Reply:
I'm not aware of a finite element whose nodes are divergences (although it would be possible to do anyway - div is just sum grad). Nodes are by definition scalar valued, so I don't think curl could come up. I think this is a kind of theoretical case...

Reply:
I think there's a slight misunderstanding in what the "derivative" means here. FIAT functionals are implemented as quadrature rules that specify for each point in the quadrature rule what the weight is. Then we ask what are we integrating, and the answer is that we integrate either (components of) an expression or (components of) derivatives of an expression. The way we say what the derivatives are is that we specify (for each spatial component of the expression) what the order of the derivative is. So I think we always just need to take `grad` (or perhaps `expr.dx(...)`).

Does that make sense?

Reply:
I think its a bit clearer now. By integrating, do you simply mean the sum of the quadrature weights multiplied by the values of the nodes fed the given expression?

Reply:
Yes. See the description of FIAT functionals here: https://github.com/FEniCS/fiat/blob/master/FIAT/functional.py#L39

A functional (a node) is a rule `f[expr]` that eats an expression and produces a number. They're almost all specified as an integral with some selected measure. For example, all of the point evaluation functionals are
```
f_p[expr] = \int_K \delta(x - p) expr dx = expr(p)
```
(for evaluation at the point `p`).

Many of the other functionals are moments against components of the expression against selected sets of polynomials.

So what FIAT does is encode this integration rule in a set of weights and points (via numerical quadrature) and so we have

```
f[expr] = \sum_q w_q expr(\phi_q)
```

With some potentially fancier stuff happening if you only care about a single component of the expression.

Reply:
... except that delta is a distribution not a function and doesn't go inside an integral, but the core idea is right.

Since the space we are interpolating into is finite dimensional, all nodes are equivalent to integrals so we could theoretically do away with derivatives at all. However, as @rckirby pointed out, that looks a bit messy so explicitly using derivatives of the expression is probably a better move.